### PR TITLE
Refactor decimal model

### PR DESCRIPTION
**Pull request by lrubasze**
## Summary
Optimize Decimal model

## Details
`Decimal` type info:
- size: `192 bits`
- decimal places: `18`
- fractional part: `60 bits`/`18 digits`
- integer part: `132 bits` /`40 digits`
- max: `3138550867693340381917894711603833208051.177722232017256447`
- min: `-3138550867693340381917894711603833208051.177722232017256448 ` 

`PreciseDecimal` type info:
- size: `256 bits`
- decimal places: `36`
- fractional part: `120 bits` / `36 digits`
- integer part: `136 bits` / `41 digits`
- max: `57896044618658097711785492504343953926634.992332820282019728792003956564819967`
- min: `-57896044618658097711785492504343953926634.992332820282019728792003956564819968`

## Testing
Respective tests have been performed

## Update Recommendations
CLI tools such as `resim`, `simulator`, `scrypto` shall be updated. 

### For dApp Developers
Scrypto code shall be revised in terms of Decimal types usage. Whether there's a threat of overflowing, since the types range has been reduced.

### CODE REVIEW

File: radix-engine-common/src/math/bnum_integer.rs [hunk 1]
```diff
@@ -120,12 +120,18 @@ macro_rules! types {
...
+    BnumI448, BInt::<7>,
     BnumI512, BInt::<8>,
     BnumI768, BInt::<12>,
+    BnumU192, BUint::<3>,
     BnumU256, BUint::<4>,
+    BnumU320, BUint::<5>,
     BnumU384, BUint::<6>,
+    BnumU448, BUint::<7>,
     BnumU512, BUint::<8>,
     BnumU768, BUint::<12>
```
Comment (iamyulong) [inline radix-engine-common/src/math/bnum_integer.rs:136]: Do you happen to notice any compile-time change due to extra types?

File: radix-engine-common/src/math/decimal.rs [hunk 2]
```diff
@@ -18,15 +19,20 @@ use crate::math::PreciseDecimal;
...
 use crate::*;
 
-/// `Decimal` represents a 256 bit representation of a fixed-scale decimal number.
+/// `Decimal` represents a 192 bit representation of a fixed-scale decimal number.
 ///
 /// The finite set of values are of the form `m / 10^18`, where `m` is
-/// an integer such that `-2^(256 - 1) <= m < 2^(256 - 1)`.
+/// an integer such that `-2^(192 - 1) <= m < 2^(192 - 1)`.
+///
+/// Fractional part: 60 bits/18 digits
```
Comment (iamyulong) [inline radix-engine-common/src/math/decimal.rs:27]: Can we add `~` because it's not exact 60 bits?

File: radix-engine-common/src/math/decimal.rs [hunk 3]
```diff
@@ -372,11 +378,11 @@ impl Mul<Decimal> for Decimal {
...
-        // Use BnumI384 (BInt<6>) to not overflow.
-        let a = BnumI384::from(self.0);
-        let b = BnumI384::from(other.0);
-        let c = a * b / BnumI384::from(Self::ONE.0);
-        let c_256 = BnumI256::try_from(c).expect("Overflow");
+        // Use BnumI256 (BInt<4>) to not overflow.
+        let a = BnumI256::from(self.0);
+        let b = BnumI256::from(other.0);
+        let c = a * b / BnumI256::from(Self::ONE.0);
+        let c_256 = BnumI192::try_from(c).expect("Overflow");
```
Comment (iamyulong) [inline radix-engine-common/src/math/decimal.rs:385]: `c_256` => `c_192`, and same below

**Review by iamyulong — APPROVED** [review]

File: radix-engine-common/src/math/decimal.rs (hunk 3 above)
Comment (lrubasze) [reply radix-engine-common/src/math/decimal.rs]: Done

File: radix-engine-common/src/math/decimal.rs (hunk 2 above)
Comment (lrubasze) [reply radix-engine-common/src/math/decimal.rs]: Done

**Pull request merged**